File: Propuesta 1/SIMION/examples/pa2text/lua/pa2text.lua
--
-- pa2text.lua
-- Reads the contents of a SIMION potential array file and writes the contents
-- to standard output.
--
-- This may be run from the command-line:
--
--   simion.exe --nogui lua pa2text.lua <file>
--
-- or simply from the "Run Lua Program" button on the SIMION main screen.
--
-- By default it loads the simple.pa# file if no PA file is specified.
--
-- D.Manura, Scientific Instrument Services, Inc.
--

assert(simion.pas, 'This example requires SIMION 8.1.')

local path = ... or 'simple.pa#'

local outpath = path .. ".txt"
local outfile = assert(io.open(outpath, "w"))

---- example reading
local pa = simion.pas:open(path)

-- print header parameters
outfile:write("begin_header\n")
local symmetry = pa.symmetry_type:gsub('^[23]d', '') -- simplify
outfile:write("symmetry=" .. symmetry .. "\n")
outfile:write("max_voltage=" .. pa.debug_max_voltage .. "\n")
outfile:write("nx=" .. pa.nx .. "\n")
outfile:write("ny=" .. pa.ny .. "\n")
outfile:write("nz=" .. pa.nz .. "\n")
outfile:write("mirror_x=" .. (pa.mirror_x and 1 or 0) .. "\n")
outfile:write("mirror_y=" .. (pa.mirror_y and 1 or 0) .. "\n")
outfile:write("mirror_z=" .. (pa.mirror_z and 1 or 0) .. "\n")
outfile:write("field=" .. pa.potential_type .. "\n")
outfile:write("ng=" .. (pa.ng or 100) .. "\n")
outfile:write("end_header\n")


outfile:write("begin_points\n")
for x,y,z in pa:points() do
   local potential, electrode = pa:point(x,y,z)
   local electrode_str = electrode and "1" or "0"
   outfile:write(electrode_str .. "," .. potential .. "\n")
end
outfile:write("end_points\n")

outfile:close()

print("finished writing ", outpath)

File: Propuesta 1/SIMION/examples/pa2text/cpp/pa2text.cpp
/**
 * @file pa2text.cpp
 * Reads the contents of a SIMION potential array file and writes the
 * contents to standard output.
 *
 * David Manura, Scientific Instrument Services, Inc.
 * $Revision$ $Date$ Created 2003-10.
 */
#include <iostream>
#include <fstream>
#include <simion/pa.h>

using namespace std;
using namespace simion;

void print_usage()
{
    cerr << "usage: pa2text <file>" << endl;
}

// note: issue on precision of double printing
int main(int argc, char** argv)
{
    ifstream is;

    if(argc < 2) {
        print_usage();
        return 1;
    }

    string filename = argv[1];

    ofstream outfile((filename + ".txt").c_str());

    PA pa;
    pa.load(filename);

    outfile
         << "file_type=" << "simion_pa" << endl;

    outfile
         << "begin_header" << endl
         << "mode=" << pa.mode() << endl
         << "symmetry=" << (pa.symmetry() == CYLINDRICAL ?
                            "cylindrical" : "PLANAR") << endl
         << "max_voltage=" << pa.max_voltage() << endl
         << "nx=" << pa.nx() << endl
         << "ny=" << pa.ny() << endl
         << "nz=" << pa.nz() << endl
         << "mirror_x=" << !!(pa.mirror_x()) << endl
         << "mirror_y=" << !!(pa.mirror_y()) << endl
         << "mirror_z=" << !!(pa.mirror_z()) << endl
         << "magnetic_pa=" << pa.field_string(pa.field_type()) << endl
         << "ng=" << pa.ng() << endl
         ;
    outfile
         << "end_header" << endl;

    // note: output in row-major order:
    //   double points[nz][ny][nz]
    outfile << "begin_points" << endl;
    
    for(int z=0; z<pa.nz(); z++) {
        for(int y=0; y<pa.ny(); y++) {
            for(int x=0; x<pa.nx(); x++) {
                PAPoint point = pa.point(x, y, z);
                outfile << point.electrode << "," << point.potential << endl;
            }
        }
    }
    outfile << "end_points" << endl;

    return 0;
}

